File: sources/entity/tests/test_sec_financials.php
<?php
/**
 * Test suite for SEC EDGAR XBRL financial data extraction.
 * Run: php tests/test_sec_financials.php
 */

require_once __DIR__ . '/../php/tools.php';

$config = json_decode(file_get_contents(__DIR__ . '/../php/settings.json'), true);
$tools = new LookupTools($config);

$passed = 0;
$failed = 0;

function test(string $label, $expected, $actual): void
{
    global $passed, $failed;
    if ($expected === $actual) {
        echo "  PASS: {$label}\n";
        $passed++;
    } else {
        echo "  FAIL: {$label}\n    Expected: " . json_encode($expected) . "\n    Got:      " . json_encode($actual) . "\n";
        $failed++;
    }
}

function testContains(string $label, string $needle, string $haystack): void
{
    global $passed, $failed;
    if (str_contains($haystack, $needle)) {
        echo "  PASS: {$label}\n";
        $passed++;
    } else {
        echo "  FAIL: {$label} — expected to contain \"{$needle}\"\n    Got: " . substr($haystack, 0, 300) . "\n";
        $failed++;
    }
}

echo "=== SEC EDGAR XBRL Financials ===\n\n";

// 1. Apple — large company, standard US-GAAP tags
echo "1. Apple (CIK 320193):\n";
$result = $tools->secEdgarFinancials('320193');
test('Returns data', true, strlen($result) > 0);
testContains('Has entity name', 'Apple', $result);
testContains('Has Revenue', 'Revenue', $result);
testContains('Has Net Income', 'Net Income', $result);
testContains('Has Total Assets', 'Total Assets', $result);
testContains('Has dollar amounts', '$', $result);
echo "\n{$result}\n\n";

// 2. Goldman Sachs — financial company, uses RevenuesNetOfInterestExpense
echo "2. Goldman Sachs (CIK 886982):\n";
$result2 = $tools->secEdgarFinancials('886982');
test('Returns data', true, strlen($result2) > 0);
testContains('Has entity name', 'Goldman Sachs', $result2);
testContains('Has Revenue', 'Revenue', $result2);
testContains('Has Total Assets', 'Total Assets', $result2);
echo "\n{$result2}\n\n";

// 3. CrowdStrike — uses IncludingAssessedTax variant
echo "3. CrowdStrike (CIK 1535527):\n";
$result3 = $tools->secEdgarFinancials('1535527');
test('Returns data', true, strlen($result3) > 0);
testContains('Has entity name', 'CROWDSTRIKE', $result3);
testContains('Has Revenue', 'Revenue', $result3);
echo "\n{$result3}\n\n";

// 4. Invalid CIK — should return empty
echo "4. Invalid CIK (0):\n";
$result4 = $tools->secEdgarFinancials('0');
test('Returns empty', '', $result4);
echo "\n";

// 5. Private company CIK — should return empty
echo "5. Non-existent CIK (9999999999):\n";
$result5 = $tools->secEdgarFinancials('9999999999');
test('Returns empty', '', $result5);
echo "\n";

echo "=== Results: {$passed} passed, {$failed} failed ===\n";
exit($failed > 0 ? 1 : 0);

File: sources/entity/app/tests/test_sec_financials.py
"""
Python port of tests/test_sec_financials.php — SEC EDGAR XBRL financials.

Faithful port. The PHP suite does NOT construct any hardcoded XBRL/JSON fixture: every
assertion runs `secEdgarFinancials(<CIK>)`, which issues a live request to
https://data.sec.gov/api/xbrl/companyfacts/CIK<padded>.json and parses/formats the reply.
So all assertions are downstream of a live call and the module is marked `live` (skipped
unless RUN_LIVE is set).

PHP secEdgarFinancials -> Python sec_edgar_financials.
"""
import os

import pytest

from config import load_config
from tools import LookupTools

pytestmark = [
    pytest.mark.live,
    pytest.mark.skipif(not os.environ.get("RUN_LIVE"), reason="live network"),
]


@pytest.fixture(scope="module")
def t():
    return LookupTools(load_config(), progress_callback=None)


# 1. Apple — large company, standard US-GAAP tags
def test_apple(t):
    result = t.sec_edgar_financials("320193")
    assert len(result) > 0, "Returns data"
    assert "Apple" in result, "Has entity name"
    assert "Revenue" in result, "Has Revenue"
    assert "Net Income" in result, "Has Net Income"
    assert "Total Assets" in result, "Has Total Assets"
    assert "$" in result, "Has dollar amounts"


# 2. Goldman Sachs — financial company, uses RevenuesNetOfInterestExpense
def test_goldman_sachs(t):
    result = t.sec_edgar_financials("886982")
    assert len(result) > 0, "Returns data"
    assert "Goldman Sachs" in result, "Has entity name"
    assert "Revenue" in result, "Has Revenue"
    assert "Total Assets" in result, "Has Total Assets"


# 3. CrowdStrike — uses IncludingAssessedTax variant
def test_crowdstrike(t):
    result = t.sec_edgar_financials("1535527")
    assert len(result) > 0, "Returns data"
    assert "CROWDSTRIKE" in result, "Has entity name"
    assert "Revenue" in result, "Has Revenue"


# 4. Invalid CIK — should return empty
def test_invalid_cik(t):
    result = t.sec_edgar_financials("0")
    assert result == "", "Returns empty"


# 5. Non-existent CIK — should return empty
def test_nonexistent_cik(t):
    result = t.sec_edgar_financials("9999999999")
    assert result == "", "Returns empty"
